File: Experimental/API/CuriosityPictures/CuriosityPictures.py
import os
import random
import requests

def rover(starting_number, max_iterations, size_threshold, num_pictures):
    save_directory = "./Experimental/API/CuriosityPictures/pictures"
    fetch_directory = "./Experimental/API/CuriosityPictures/pictures/fetch"

    
    params = {"earth_date": "2016-10-17", "api_key": "7QVxWTvL4YZkPFPAgWQGydwacGjiVFoCN8UdnOcU"}
    url = "https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos"
    response = requests.get(url, params=params)
    data = response.json()

    if "photos" in data:
        photos = data["photos"]
        unique_images = set()
        
        iterations_done = 0
        pictures_downloaded = 0
        
        while pictures_downloaded < num_pictures and iterations_done < max_iterations:
            if starting_number in unique_images:
                break
            
            random_photo = random.choice(photos)
            img_src = random_photo["img_src"]
            image_data = requests.get(img_src).content
            
            if len(image_data) >= size_threshold:
                image_hash = hash(image_data)  # Calculate hash to check uniqueness
                image_name = f"image{starting_number}.jpg"
                
                # Save to the fetch directory
                fetch_save_path = os.path.join(fetch_directory, f"image{pictures_downloaded}.jpg")
                
                # If the image is unique, save it in both directories
                if image_hash not in unique_images:
                    unique_images.add(image_hash)
                    
                    with open(fetch_save_path, "wb") as f:
                        f.write(image_data)
                        print(f"Fetched {image_name}")
                    
                    # main_save_path = os.path.join(save_directory, image_name)
                    # with open(main_save_path, "wb") as f:
                    #     f.write(image_data)
                    #     print(f"Saved {image_name} in main directory")
                    
                    pictures_downloaded += 1
            
            iterations_done += 1
            starting_number += 1

# Set your desired values for starting_number, max_iterations, size_threshold, and num_pictures
starting_number = 89000
max_iterations = 100
size_threshold = 50000  # Example size in bytes
num_pictures = 5

rover(starting_number, max_iterations, size_threshold, num_pictures)

File: Experimental/API/JS/curiosityPictures.js
const fs = require('fs');
const path = require('path');
const fetch = require('node-fetch');
const crypto = require('crypto');

async function rover(starting_number, max_iterations, size_threshold, num_pictures) {
    const save_directory = "./Experimental/API/JS/JSPictures";
    const fetch_directory = "./Experimental/API/JS/JSPictures";

    const params = { "earth_date": "2016-10-17", "api_key": "7QVxWTvL4YZkPFPAgWQGydwacGjiVFoCN8UdnOcU" };
    const url = "https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos";

    try {
        const response = await fetch(`${url}?${new URLSearchParams(params)}`);
        const data = await response.json();

        if ("photos" in data) {
            const photos = data.photos;
            const uniqueImages = new Set();

            let iterationsDone = 0;
            let picturesDownloaded = 0;

            while (picturesDownloaded < num_pictures && iterationsDone < max_iterations) {
                if (uniqueImages.has(starting_number)) {
                    break;
                }

                const randomPhoto = photos[Math.floor(Math.random() * photos.length)];
                const imgSrc = randomPhoto.img_src;
                const imageResponse = await fetch(imgSrc);
                const imageBuffer = await imageResponse.buffer();

                if (imageBuffer.length >= size_threshold) {
                    const imageHash = crypto.createHash('md5').update(imageBuffer).digest('hex');
                    const imageFileName = `image${starting_number}.jpg`;

                    // Save to the fetch directory
                    const fetchSavePath = path.join(fetch_directory, `image${picturesDownloaded}.jpg`);

                    // If the image is unique, save it in both directories
                    if (!uniqueImages.has(imageHash)) {
                        uniqueImages.add(imageHash);

                        await fs.promises.writeFile(fetchSavePath, imageBuffer);
                        console.log(`Fetched ${imageFileName}`);

                        // const mainSavePath = path.join(save_directory, imageFileName);
                        // await fs.promises.writeFile(mainSavePath, imageBuffer);
                        // console.log(`Saved ${imageFileName} in main directory`);

                        picturesDownloaded += 1;
                    }
                }

                iterationsDone += 1;
                starting_number += 1;
            }
        }
    } catch (error) {
        console.error("An error occurred:", error);
    }
}

// Set your desired values for starting_number, max_iterations, size_threshold, and num_pictures
const starting_number = 89000;
const max_iterations = 100;
const size_threshold = 50000; // Example size in bytes
const num_pictures = 5;

rover(starting_number, max_iterations, size_threshold, num_pictures);
